master/internal/stream/system.go

```diff
@@ -255,72 +233,75 @@ func (ps *PublisherSet) Websocket(
 		}
 	}()
 
-	// reads is where we collect SubscriptionModMsg messages we read from the websocket until
-	// waitForSomething() delivers those messages to the websocket goroutine.
-	var reads []SubscriptionModMsg
-
-	// always be reading for new subscriptions
+	// monitor for new StartupMsgs
+	var restartMsg StartupMsg
+	var restartRequired bool
 	go func() {
 		defer streamer.Close()
 		for {
-			var mods SubscriptionModMsg
-			err := socket.ReadJSON(&mods)
+			var newStartup StartupMsg
+			err := socket.ReadJSON(&newStartup)
```

Comment:
Imagine that somebody sends two startup messages, one right after the other.  What is the proper server behavior in that case?

I can tell you what server behavior I would expect as the client.  I would expect that for every startup message I send, I would like to see the full set of startup messages and then a sync message.  That sync message is useful for things like populating a list view in a webui.  I don't want to populate the view as trials come in off the wire, I want to compile the whole list while showing a loading spinner, then populate the whole view in one atomic change.  That's only possible if the server sends an explicit sync.  I think I left a TODO in the prototype where I thought such a sync would be needed.

Then, it would sure be nice if the server processed my messages one-at-a-time.  I can pipeline things if I want, and can expect that as long as the connection doesn't break, 1 startup message corresponds to one sync message.  That's easy to think about.

But this pattern of saving a pointer and using a dirty flag doesn't cut it for that situation.  For that situation, you really need to stop reading from the websocket as soon as you have read one message.  I think with a more aggressive reorg of this function you could model it just about right, where also your go structs are more one-time-use and there's less of resetting state going on.